src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,479 @@
+ *   This library is free software; you can redistribute it and/or         *
+ *   modify it under the terms of the GNU Library General Public           *
+ *   License as published by the Free Software Foundation; either          *
+ *   version 2 of the License, or (at your option) any later version.      *
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
+# include <cfloat>
+
+# include <Inventor/nodes/SoDepthBuffer.h>
+# include <Inventor/nodes/SoDrawStyle.h>
+# include <Inventor/nodes/SoLineSet.h>
+# include <Inventor/nodes/SoMaterial.h>
+# include <Inventor/nodes/SoPickStyle.h>
+# include <Inventor/nodes/SoVertexProperty.h>
+# include <Inventor/nodes/SoSeparator.h>
+# include <Inventor/nodes/SoBaseColor.h>
+# include <Inventor/SbVec3f.h>
+#endif
+
+#include <Base/Parameter.h>
+#include <App/Application.h>
+#include <Mod/Part/App/PropertyTopoShape.h>
+
+#include <Gui/Application.h>
+#include <Gui/Document.h>
+#include <Gui/MDIView.h>
+#include <Gui/View3DInventor.h>
+#include <Gui/View3DInventorViewer.h>
+#include <Gui/ViewProvider.h>
+#include <Gui/ViewProviderDocumentObject.h>
+
+#include <Gui/SoFCBoundingBox.h>
+
+#include "ViewProviderGridExtension.h"
+
+
+using namespace PartGui;
+using namespace std;
+
+EXTENSION_PROPERTY_SOURCE(PartGui::ViewProviderGridExtension, Gui::ViewProviderExtension)
+
+namespace PartGui {
+
+class GridExtensionP : public ParameterGrp::ObserverType {
+public:
+    explicit GridExtensionP(ViewProviderGridExtension *);
+    ~GridExtensionP();
+
+    void drawGrid(bool cameraUpdate = false);
+
+    void setEnabled(bool enable);
+    bool getEnabled();
+
+    SoSeparator * getGridRoot();
+
+    // Configurable parameters (to be configured by specific VP)
+    int GridSizePixelThreshold = 15;
+    int GridNumberSubdivision = 10;
+    int GridLinePattern = 0x0f0f;
+    int GridDivLinePattern = 0xffff;
+    int GridLineWidth = 1;
+    int GridDivLineWidth = 2;
+    unsigned int GridLineColor;
+    unsigned int GridDivLineColor;
```

Comment:
Ultimately the color will be used as a SoBaseColor. 

On one hand, while ViewProviderGridExtension is (and must be) agnostic about how it receives the colors, it is difficult to obviate the fact that colors are stored as parameters in FreeCAD as unsigned ints. So any ViewProvider trying to implement a grid will most likely have these preferences in unsigned int format. This means that using an intermediate format (App:Color) will incur a penalty (it stores the color as floats).

On the other hand, using App::Color is not meritless at all. It has the advantage of leveraging a much stronger type check (than a generic unsigned int). 

As we are implementing parameters the observer way to avoid polling, I tend to side with you that having a formal API with stronger type checking outweighs the performance penalty.

Will mark for change after we resolve the other issues.